File: backend/Services/SnapTradeService.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using StockAnalyzer.Models;

namespace StockAnalyzer.Services;

public class SnapTradeService : ISnapTradeService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;
    private readonly ILogger<SnapTradeService> _logger;

    public SnapTradeService(
        HttpClient httpClient,
        IConfiguration configuration,
        ILogger<SnapTradeService> logger)
    {
        _httpClient = httpClient;
        _configuration = configuration;
        _logger = logger;
    }

    public async Task<SnapTradeUser> CreateUserAsync(string userId)
    {
        try
        {
            var body = new { userId = userId };
            var response = await _httpClient.PostAsJsonAsync("/snapTrade/v1/register", body);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            return new SnapTradeUser
            {
                Id = result.GetProperty("id").GetString() ?? "",
                UserId = userId,
                Email = result.TryGetProperty("email", out var email) ? email.GetString() : null,
                CreatedAt = result.TryGetProperty("createdAt", out var createdAt) 
                    ? createdAt.GetDateTime() 
                    : DateTime.UtcNow
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating SnapTrade user for userId: {UserId}", userId);
            throw;
        }
    }

    public async Task<SnapTradeUser?> GetUserAsync(string userId, string userSecret)
    {
        try
        {
            var queryParams = $"?userId={Uri.EscapeDataString(userId)}&userSecret={Uri.EscapeDataString(userSecret)}";
            var response = await _httpClient.GetAsync($"/snapTrade/v1/users{queryParams}");
            
            if (response.IsSuccessStatusCode)
            {
                var result = await response.Content.ReadFromJsonAsync<JsonElement>();
                return new SnapTradeUser
                {
                    Id = result.GetProperty("id").GetString() ?? "",
                    UserId = userId,
                    Email = result.TryGetProperty("email", out var email) ? email.GetString() : null
                };
            }
            return null;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching SnapTrade user for userId: {UserId}", userId);
            throw;
        }
    }

    public async Task<List<Brokerage>> GetBrokeragesAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync("/snapTrade/v1/brokerages");
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            var brokerages = new List<Brokerage>();

            if (result.TryGetProperty("brokerages", out var brokeragesElement))
            {
                foreach (var brokerage in brokeragesElement.EnumerateArray())
                {
                    brokerages.Add(new Brokerage
                    {
                        Id = brokerage.GetProperty("id").GetString() ?? "",
                        Name = brokerage.GetProperty("name").GetString() ?? "",
                        DisplayName = brokerage.TryGetProperty("displayName", out var dn) ? dn.GetString() : null,
                        SupportsOAuth = brokerage.TryGetProperty("supportsOAuth", out var oauth) && oauth.GetBoolean()
                    });
                }
            }

            return brokerages;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching brokerages");
            throw;
        }
    }

    public async Task<string> InitiateConnectionAsync(string userId, string userSecret, string redirectUri)
    {
        try
        {
            var body = new
            {
                userId = userId,
                userSecret = userSecret,
                redirectUri = redirectUri
            };

            var response = await _httpClient.PostAsJsonAsync("/snapTrade/v1/auth/login", body);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            return result.TryGetProperty("loginLink", out var loginLink)
                ? loginLink.GetString() ?? ""
                : result.TryGetProperty("redirectUri", out var redirect)
                    ? redirect.GetString() ?? ""
                    : "";
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error initiating connection for userId: {UserId}", userId);
            throw;
        }
    }

    public async Task<List<Account>> GetAccountsAsync(string userId, string userSecret)
    {
        try
        {
            var queryParams = $"?userId={Uri.EscapeDataString(userId)}&userSecret={Uri.EscapeDataString(userSecret)}";
            var response = await _httpClient.GetAsync($"/snapTrade/v1/accounts{queryParams}");
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            var accounts = new List<Account>();

            if (result.TryGetProperty("accounts", out var accountsElement))
            {
                foreach (var account in accountsElement.EnumerateArray())
                {
                    accounts.Add(ParseAccount(account));
                }
            }
            else if (result.ValueKind == JsonValueKind.Array)
            {
                foreach (var account in result.EnumerateArray())
                {
                    accounts.Add(ParseAccount(account));
                }
            }

            return accounts;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching accounts for userId: {UserId}", userId);
            throw;
        }
    }

    public async Task<Account?> GetAccountAsync(string userId, string userSecret, string accountId)
    {
        try
        {
            var queryParams = $"?userId={Uri.EscapeDataString(userId)}&userSecret={Uri.EscapeDataString(userSecret)}";
            var response = await _httpClient.GetAsync($"/snapTrade/v1/accounts/{accountId}{queryParams}");
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            return ParseAccount(result);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching account {AccountId} for userId: {UserId}", accountId, userId);
            throw;
        }
    }

    public async Task<AccountBalance> GetAccountBalanceAsync(string userId, string userSecret, string accountId)
    {
        try
        {
            var queryParams = $"?userId={Uri.EscapeDataString(userId)}&userSecret={Uri.EscapeDataString(userSecret)}";
            var response = await _httpClient.GetAsync($"/snapTrade/v1/accounts/{accountId}/balances{queryParams}");
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            return new AccountBalance
            {
                AccountId = accountId,
                TotalCash = result.TryGetProperty("totalCash", out var cash) ? cash.GetDecimal() : 0,
                BuyingPower = result.TryGetProperty("buyingPower", out var buying) ? buying.GetDecimal() : 0,
                Currency = result.TryGetProperty("currency", out var currency) ? currency.GetString() ?? "USD" : "USD",
                Positions = result.TryGetProperty("positions", out var positions) ? positions.GetArrayLength() : 0
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching account balance for account {AccountId}", accountId);
            throw;
        }
    }

    public async Task<List<Holding>> GetAccountHoldingsAsync(string userId, string userSecret, string accountId)
    {
        try
        {
            var queryParams = $"?userId={Uri.EscapeDataString(userId)}&userSecret={Uri.EscapeDataString(userSecret)}";
            var response = await _httpClient.GetAsync($"/snapTrade/v1/accounts/{accountId}/holdings{queryParams}");
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            var holdings = new List<Holding>();

            if (result.TryGetProperty("holdings", out var holdingsElement))
            {
                foreach (var holding in holdingsElement.EnumerateArray())
                {
                    holdings.Add(ParseHolding(holding));
                }
            }
            else if (result.TryGetProperty("positions", out var positionsElement))
            {
                foreach (var holding in positionsElement.EnumerateArray())
                {
                    holdings.Add(ParseHolding(holding));
                }
            }

            return holdings;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching holdings for account {AccountId}", accountId);
            throw;
        }
    }

    public async Task<Portfolio> GetPortfolioAsync(string userId, string userSecret)
    {
        var accounts = await GetAccountsAsync(userId, userSecret);
        
        // Load holdings for each account
        foreach (var account in accounts)
        {
            account.Holdings = await GetAccountHoldingsAsync(userId, userSecret, account.Id);
        }

        var totalBalance = accounts.Sum(a => a.Balance ?? 0);
        var totalGainLoss = accounts.Sum(a => a.Holdings.Sum(h => h.GainLoss));
        var totalGainLossPercent = totalBalance > 0 
            ? (totalGainLoss / (totalBalance - totalGainLoss)) * 100 
            : 0;

        return new Portfolio
        {
            UserId = userId,
            Accounts = accounts,
            TotalBalance = totalBalance,
            TotalGainLoss = totalGainLoss,
            TotalGainLossPercent = totalGainLossPercent,
            Currency = accounts.FirstOrDefault()?.Currency ?? "USD"
        };
    }

    public async Task<TradeExecution> PlaceTradeAsync(string userId, string userSecret, string accountId, TradeOrder order)
    {
        try
        {
            var body = new
            {
                userId = userId,
                userSecret = userSecret,
                accountId = accountId,
                symbol = order.Symbol,
                action = order.Action,
                units = order.Quantity,
                orderType = order.OrderType,
                timeInForce = order.TimeInForce,
                price = order.LimitPrice,
                stop = order.StopPrice
            };

            var response = await _httpClient.PostAsJsonAsync("/snapTrade/v1/trade/place", body);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            return ParseTradeExecution(result);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error placing trade for account {AccountId}", accountId);
            throw;
        }
    }

    public async Task<TradeExecution?> GetTradeStatusAsync(string userId, string userSecret, string accountId, string tradeId)
    {
        try
        {
            var queryParams = $"?userId={Uri.EscapeDataString(userId)}&userSecret={Uri.EscapeDataString(userSecret)}";
            var response = await _httpClient.GetAsync($"/snapTrade/v1/trade/{tradeId}{queryParams}");
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            return ParseTradeExecution(result);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching trade status for trade {TradeId}", tradeId);
            throw;
        }
    }

    private Account ParseAccount(JsonElement element)
    {
        return new Account
        {
            Id = element.GetProperty("id").GetString() ?? "",
            Name = element.TryGetProperty("name", out var name) ? name.GetString() ?? "" : "",
            AccountNumber = element.TryGetProperty("accountNumber", out var accNum) ? accNum.GetString() ?? "" : "",
            Type = element.TryGetProperty("type", out var type) ? type.GetString() ?? "" : "",
            BrokerageId = element.TryGetProperty("brokerageId", out var brokerId) ? brokerId.GetString() ?? "" : "",
            Balance = element.TryGetProperty("balance", out var balance) ? balance.GetDecimal() : null,
            Currency = element.TryGetProperty("currency", out var currency) ? currency.GetString() ?? "USD" : "USD"
        };
    }

    private Holding ParseHolding(JsonElement element)
    {
        var quantity = element.TryGetProperty("quantity", out var qty) ? qty.GetDecimal() : 0;
        var avgPrice = element.TryGetProperty("averagePurchasePrice", out var avg) ? avg.GetDecimal() : 0;
        var currentPrice = element.TryGetProperty("currentPrice", out var curr) ? curr.GetDecimal() : 0;
        var totalValue = element.TryGetProperty("totalValue", out var total) ? total.GetDecimal() : (quantity * currentPrice);
        var gainLoss = totalValue - (quantity * avgPrice);
        var gainLossPercent = (quantity * avgPrice) > 0 ? (gainLoss / (quantity * avgPrice)) * 100 : 0;

        return new Holding
        {
            Symbol = element.TryGetProperty("symbol", out var symbol) ? symbol.GetString() ?? "" : "",
            Quantity = quantity,
            AveragePurchasePrice = avgPrice,
            CurrentPrice = currentPrice,
            TotalValue = totalValue,
            GainLoss = gainLoss,
            GainLossPercent = gainLossPercent,
            Currency = element.TryGetProperty("currency", out var currency) ? currency.GetString() ?? "USD" : "USD"
        };
    }

    private TradeExecution ParseTradeExecution(JsonElement element)
    {
        return new TradeExecution
        {
            Id = element.TryGetProperty("id", out var id) ? id.GetString() ?? "" : "",
            AccountId = element.TryGetProperty("accountId", out var accId) ? accId.GetString() ?? "" : "",
            Symbol = element.TryGetProperty("symbol", out var symbol) ? symbol.GetString() ?? "" : "",
            Action = element.TryGetProperty("action", out var action) ? action.GetString() ?? "" : "",
            Quantity = element.TryGetProperty("quantity", out var qty) ? qty.GetDecimal() : 0,
            Price = element.TryGetProperty("price", out var price) ? price.GetDecimal() : 0,
            Status = element.TryGetProperty("status", out var status) ? status.GetString() ?? "" : "",
            ExecutedAt = element.TryGetProperty("executedAt", out var exec) 
                ? exec.GetDateTime() 
                : DateTime.UtcNow
        };
    }
}

File: backend_python/services/snaptrade_service.py
"""SnapTrade API client for broker linking and portfolio data."""
import logging

import httpx

from config import SNAPTRADE_API_URL, SNAPTRADE_CLIENT_ID, SNAPTRADE_CONSUMER_KEY
from models.snaptrade_models import Account, Brokerage, Holding, Portfolio, SnapTradeUser

logger = logging.getLogger(__name__)

_headers = {
    "X-API-Key": SNAPTRADE_CLIENT_ID,
    "X-Consumer-Key": SNAPTRADE_CONSUMER_KEY,
    "Content-Type": "application/json",
}


async def _aget(path: str, params: dict | None = None) -> dict:
    url = f"{SNAPTRADE_API_URL.rstrip('/')}{path}"
    async with httpx.AsyncClient() as client:
        resp = await client.get(url, headers=_headers, params=params, timeout=30.0)
        resp.raise_for_status()
        return resp.json()


async def _apost(path: str, json: dict) -> dict:
    url = f"{SNAPTRADE_API_URL.rstrip('/')}{path}"
    async with httpx.AsyncClient() as client:
        resp = await client.post(url, headers=_headers, json=json, timeout=30.0)
        resp.raise_for_status()
        return resp.json()


async def create_user(user_id: str) -> SnapTradeUser:
    result = await _apost("/snapTrade/v1/register", {"userId": user_id})
    return SnapTradeUser(
        id=result.get("id", ""),
        user_id=user_id,
        email=result.get("email"),
        created_at=result.get("createdAt"),
    )


async def get_brokerages() -> list[Brokerage]:
    result = await _aget("/snapTrade/v1/brokerages")
    brokerages = []
    for b in result.get("brokerages", []):
        brokerages.append(
            Brokerage(
                id=b.get("id", ""),
                name=b.get("name", ""),
                display_name=b.get("displayName"),
                supports_oauth=b.get("supportsOAuth", False),
            )
        )
    return brokerages


async def initiate_connection(user_id: str, user_secret: str, redirect_uri: str) -> str:
    result = await _apost(
        "/snapTrade/v1/auth/login",
        {"userId": user_id, "userSecret": user_secret, "redirectUri": redirect_uri},
    )
    return result.get("loginLink") or result.get("redirectUri") or ""


def _parse_account(element: dict) -> Account:
    return Account(
        id=element.get("id", ""),
        name=element.get("name", ""),
        account_number=element.get("accountNumber", ""),
        type=element.get("type", ""),
        brokerage_id=element.get("brokerageId", ""),
        balance=element.get("balance"),
        currency=element.get("currency", "USD"),
    )


def _parse_holding(element: dict) -> Holding:
    qty = float(element.get("quantity", 0) or 0)
    avg = float(element.get("averagePurchasePrice", 0) or 0)
    curr = float(element.get("currentPrice", 0) or 0)
    total = float(element.get("totalValue", 0) or (qty * curr))
    gain_loss = total - (qty * avg)
    gain_pct = (gain_loss / (qty * avg) * 100) if (qty * avg) else 0
    return Holding(
        symbol=element.get("symbol", ""),
        quantity=qty,
        average_purchase_price=avg,
        current_price=curr,
        total_value=total,
        gain_loss=gain_loss,
        gain_loss_percent=gain_pct,
        currency=element.get("currency", "USD"),
    )


async def get_accounts(user_id: str, user_secret: str) -> list[Account]:
    result = await _aget(
        "/snapTrade/v1/accounts",
        {"userId": user_id, "userSecret": user_secret},
    )
    accounts = []
    raw = result.get("accounts")
    if isinstance(raw, list):
        for acc in raw:
            accounts.append(_parse_account(acc))
    elif isinstance(result, list):
        for acc in result:
            accounts.append(_parse_account(acc))
    return accounts


async def get_account_holdings(
    user_id: str, user_secret: str, account_id: str
) -> list[Holding]:
    result = await _aget(
        f"/snapTrade/v1/accounts/{account_id}/holdings",
        {"userId": user_id, "userSecret": user_secret},
    )
    holdings = []
    for elem in result.get("holdings", []) or result.get("positions", []):
        holdings.append(_parse_holding(elem))
    return holdings


async def get_portfolio(user_id: str, user_secret: str) -> Portfolio:
    accounts = await get_accounts(user_id, user_secret)
    for acc in accounts:
        acc.holdings = await get_account_holdings(user_id, user_secret, acc.id)
    total_balance = sum(a.balance or 0 for a in accounts)
    total_gain_loss = sum(sum(h.gain_loss for h in a.holdings) for a in accounts)
    total_gain_loss_percent = (
        (total_gain_loss / (total_balance - total_gain_loss) * 100)
        if (total_balance - total_gain_loss)
        else 0
    )
    currency = accounts[0].currency if accounts else "USD"
    return Portfolio(
        user_id=user_id,
        accounts=accounts,
        total_balance=total_balance,
        total_gain_loss=total_gain_loss,
        total_gain_loss_percent=total_gain_loss_percent,
        currency=currency,
    )
